### PR TITLE
Feature/2607 get readmetxt from fss and add to exchange set

### PR DESCRIPTION
**Pull request by Avani13709**
Code changes
Unit Test Cases
Bug fix #7487 requestedProductsNotInExchangeSet should be []
Code changes for #7485 Ensure lower camel case used for attribute names

### CODE REVIEW

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Common/Helpers/FileShareService.cs [hunk 1]
```diff
@@ -208,5 +209,74 @@ public string GenerateQueryForFss(List<Products> products)
...
+            {
+                SearchBatchResponse searchBatchResponse = await SearchBatchResponse(httpResponse);
+                if (searchBatchResponse.Entries.Count > 0)
+                {
+                    var batchResult = searchBatchResponse.Entries.FirstOrDefault();
+                    filePath =  batchResult.Files.FirstOrDefault().Links.Get.Href;                  
+                }
+                else
+                    logger.LogInformation(EventIds.ReadMeTextFileNotFound.ToEventId(), "Readme.txt file not found for {BatchId}", batchId);              
+            }
```
Comment (janvi13603) [inline UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Common/Helpers/FileShareService.cs:266]: can we log non ok response in else like below
               else
                {
                    logger.LogInformation(EventIds.QueryFileShareServiceNonOkResponse.ToEventId(), "Query File share service for readme file with uri {RequestUri} responded with {StatusCode} for {BatchId}", httpResponse.RequestMessage.RequestUri, httpResponse.StatusCode,batchId);
                }

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Common.UnitTests/Helpers/FileShareServiceTests.cs [hunk 2]
```diff
@@ -213,5 +226,86 @@ public async Task WhenGetBatchInfoBasedOnProducts_ThenReturnsSearchBatchResponse
         }
 
         #endregion GetBatchInfoBasedOnProducts
+
+        #region SearchReadMeFilePath
+        [Test]
+        public async Task WhenSearchReadMeFilePath_ThenSearchReadMeFilePathReturnsEmptyStringResponse()
```
Comment (janvi13603) [inline UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Common.UnitTests/Helpers/FileShareServiceTests.cs:232]: Kindly change WhenSearchReadMeFilePath_ThenSearchReadMeFilePathReturnsEmptyStringResponse to
WhenInvalidRequest_ThenSearchReadMeFileReturnsEmptyFilePath()

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Common.UnitTests/Helpers/FileShareServiceTests.cs [hunk 3]
```diff
@@ -213,5 +226,86 @@ public async Task WhenGetBatchInfoBasedOnProducts_ThenReturnsSearchBatchResponse
...
+            A.CallTo(() => fakeAuthTokenProvider.GetManagedIdentityAuthAsync(A<string>.Ignored)).Returns(GetFakeToken());
+            A.CallTo(() => fakeFileShareServiceClient.CallFileShareServiceApi(A<HttpMethod>.Ignored, A<string>.Ignored, A<string>.Ignored, A<string>.Ignored))
+                 .Returns(new HttpResponseMessage() { StatusCode = HttpStatusCode.BadRequest, RequestMessage = new HttpRequestMessage() { RequestUri = new Uri("http://test.com") }, Content = new StreamContent(new MemoryStream(Encoding.UTF8.GetBytes("Bad request"))) });
+
+            var response = await fileShareService.SearchReadMeFilePath(string.Empty);
+            Assert.AreEqual(string.Empty, response);
+        
+        }
+        [Test]
+        public async Task WhenSearchReadMeFilePath_ThenSearchReadMeFilePathReturnsReadMeFilePathResponse()
```
Comment (janvi13603) [inline UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Common.UnitTests/Helpers/FileShareServiceTests.cs:243]: Kindly change WhenSearchReadMeFilePath_ThenSearchReadMeFilePathReturnsReadMeFilePathResponse() to WhenValidRequest_ThenSearchReadMeFileReturnsValidFilePath()

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Common/Helpers/FileShareService.cs (hunk 1 above)
Comment (Avani13709) [reply UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Common/Helpers/FileShareService.cs:266]: Done

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Common.UnitTests/Helpers/FileShareServiceTests.cs (hunk 2 above)
Comment (Avani13709) [reply UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Common.UnitTests/Helpers/FileShareServiceTests.cs]: Done

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Common.UnitTests/Helpers/FileShareServiceTests.cs (hunk 3 above)
Comment (Avani13709) [reply UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Common.UnitTests/Helpers/FileShareServiceTests.cs]: Done

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Common.UnitTests/Helpers/FileShareServiceTests.cs [hunk 4]
```diff
@@ -213,5 +226,86 @@ public async Task WhenGetBatchInfoBasedOnProducts_ThenReturnsSearchBatchResponse
...
+            string expectedReadMeFilePath = @"batch/a07537ff-ffa2-4565-8f0e-96e61e70a9fc/files/README.TXT";
+            Assert.IsNotNull(response);
+            Assert.AreEqual(expectedReadMeFilePath, searchReadMeFileName);
+        }
+        #endregion SearchReadMeFilePath
+
+        #region DownloadReadMeFile
+
+        [Test]
+        public async Task WhenDownloadReadMeFile_ThenDownloadReadMeFileReturnsTrueResponse()
```
[the remark this answers is not in the document]
Comment (Avani13709) [reply UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Common.UnitTests/Helpers/FileShareServiceTests.cs]: Done

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Common/Helpers/FileShareService.cs [hunk 5]
```diff
@@ -247,6 +247,75 @@ private static async Task CopyFileToFolder(HttpResponseMessage httpResponse, str
...
+            else
+                logger.LogInformation(EventIds.QueryFileShareServiceNonOkResponse.ToEventId(), "Query File share service for readme file with uri {RequestUri} responded with {StatusCode} for {BatchId}", httpResponse.RequestMessage.RequestUri, httpResponse.StatusCode, batchId);
+
+            return filePath;
+        }
+
+        private static string GetLine(string filePath, int line)
+        {
+            string secondLine = string.Empty;
```
Comment (martyn-fewtrell) [inline UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Common/Helpers/FileShareService.cs:311]: Would this variable be better names as lineToFind (or something similar)?

I realise in this case we are looking for the second line but in reality this is a generic method to find the line of the number passed in.

Comment (martyn-fewtrell) [reply UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Common/Helpers/FileShareService.cs:311]: or perhaps lineFound?

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.FulfilmentService/appsettings.json [hunk 6]
```diff
@@ -11,8 +11,8 @@
   "ESSFulfilmentStorageConfiguration": {
     "StorageAccountName": "",
     "StorageAccountKey": "",
-    "StorageContainerName": "ess-fulfilment",
-    "QueueName": "ess-fulfilment-requests"
+    "StorageContainerName": "",
```
Comment (janvi13603) [inline UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.FulfilmentService/appsettings.json:14]: Kindly add StorageContainerName and QueueName

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.FulfilmentService/Services/FulfilmentDataService.cs [hunk 7]
```diff
@@ -74,5 +74,20 @@ public async Task QueryAndDownloadFileShareServiceFiles(SalesCatalogueServiceRes
...
             }
         }
+        private async Task CreateAncillaryFiles(string batchId, string exchangeSetRootPath)
+        {
+            await DownloadReadMeFile(batchId, exchangeSetRootPath);
+        }
+
+        public async Task DownloadReadMeFile(string batchId, string exchangeSetRootPath)
+        {
+            logger.LogInformation(EventIds.SearchDownloadReadMeFileRequestStart.ToEventId(), "Search and download ReadMe Text File start for {BatchId}", batchId);
```
Comment (janvi13603) [inline UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.FulfilmentService/Services/FulfilmentDataService.cs:84]: Can we log Search and Download ReadMeFileRequestStart and ReadMeFileRequestCompleted separately.

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Webjob.UnitTests/Services/FulfilmentFileShareServiceTest.cs [hunk 8]
```diff
@@ -100,5 +101,58 @@ public void WhenInValidRequest_ThenDownloadFileShareServiceFilesReturnsNoFile()
...
+        public async Task WhenRequestSearchReadMeFilePath_ThenReturnsFilePath()
+        {
+            string batchId = "7b4cdf10-adfa-4ed6-b2fe-d1543d8b7272";
+            string exchangeSetRootPath = @"D:\\Downloads";
+
+            A.CallTo(() => fakefileShareService.SearchReadMeFilePath(A<string>.Ignored)).Returns(exchangeSetRootPath);
+            var result = await fulfilmentFileShareService.SearchReadMeFilePath(batchId);
+
+            Assert.IsNotNull(result);
+            Assert.AreEqual("ReadMe Text file path found", "ReadMe Text file path found");
```
Comment (martyn-fewtrell) [inline UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Webjob.UnitTests/Services/FulfilmentFileShareServiceTest.cs:115]: What is this line doing?

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Webjob.UnitTests/Services/FulfilmentFileShareServiceTest.cs [hunk 9]
```diff
@@ -100,5 +101,58 @@ public void WhenInValidRequest_ThenDownloadFileShareServiceFilesReturnsNoFile()
             var result = fulfilmentFileShareService.DownloadFileShareServiceFiles(message, fulfilmentDataResponses, "");
             Assert.IsNotNull(result);
         }
+
+        [Test]
+        public async Task WhenRequestSearchReadMeFilePath_ThenReturnsFilePath()
```
Comment (janvi13603) [inline UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Webjob.UnitTests/Services/FulfilmentFileShareServiceTest.cs:106]: Kindly change WhenRequestSearchReadMeFilePath_ThenReturnsFilePath() to WhenValidSearchReadMeFileRequest_ThenReturnFilePath() and change WhenRequestSearchReadMeFilePath_ThenReturnsEmptyStringResponse() to WhenInvalidSearchReadMeFileRequest_ThenReturnEmptyFilePath()

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Webjob.UnitTests/Services/FulfilmentFileShareServiceTest.cs [hunk 10]
```diff
@@ -100,5 +101,58 @@ public void WhenInValidRequest_ThenDownloadFileShareServiceFilesReturnsNoFile()
...
+            string exchangeSetRootPath = @"D:\\Downloads";
+
+            A.CallTo(() => fakefileShareService.SearchReadMeFilePath(A<string>.Ignored)).Returns(exchangeSetRootPath);
+            var result = await fulfilmentFileShareService.SearchReadMeFilePath(batchId);
+
+            Assert.IsNotNull(result);
+            Assert.AreEqual("ReadMe Text file path found", "ReadMe Text file path found");
+        }
+         [Test]
+        public async Task WhenRequestSearchReadMeFilePath_ThenReturnsEmptyStringResponse()
```
Comment (martyn-fewtrell) [inline UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Webjob.UnitTests/Services/FulfilmentFileShareServiceTest.cs:118]: Is the naming on this test correct? 

Should the first part of the name infer the reason for the test as I'm not sure that it currently does?

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Webjob.UnitTests/Services/FulfilmentFileShareServiceTest.cs (hunk 8 above)
Comment (janvi13603) [reply UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Webjob.UnitTests/Services/FulfilmentFileShareServiceTest.cs:115]: This assert statement is incorrect. It should be Assert.AreEqual("ReadMe Text file path found", result); Correct if I am wrong

Comment (martyn-fewtrell) [reply UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Webjob.UnitTests/Services/FulfilmentFileShareServiceTest.cs:115]: Thanks Janvi

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Common/Helpers/FileShareService.cs (hunk 5 above)
Comment (Avani13709) [reply UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Common/Helpers/FileShareService.cs:311]: Done. Modified method.

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.FulfilmentService/Services/FulfilmentDataService.cs (hunk 7 above)
Comment (Avani13709) [reply UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.FulfilmentService/Services/FulfilmentDataService.cs]: Done

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Webjob.UnitTests/Services/FulfilmentFileShareServiceTest.cs (hunk 8 above)
Comment (Avani13709) [reply UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Webjob.UnitTests/Services/FulfilmentFileShareServiceTest.cs]: Removed Assert.AreEqual as it is not needed. Modified method.

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Webjob.UnitTests/Services/FulfilmentFileShareServiceTest.cs (hunk 9 above)
Comment (Avani13709) [reply UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Webjob.UnitTests/Services/FulfilmentFileShareServiceTest.cs]: Done

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Webjob.UnitTests/Services/FulfilmentFileShareServiceTest.cs (hunk 10 above)
Comment (Avani13709) [reply UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Webjob.UnitTests/Services/FulfilmentFileShareServiceTest.cs]: Modified.

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.FulfilmentService/appsettings.json (hunk 6 above)
Comment (Avani13709) [reply UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.FulfilmentService/appsettings.json]: done

**Review by janvi13603 — APPROVED** [review]
Changes looks good. Thanks for fixing review comments.

**Review by martyn-fewtrell — COMMENTED** [review]
Just two comment to discuss

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Common/Helpers/FileShareService.cs [hunk 11]
```diff
@@ -247,6 +247,75 @@ private static async Task CopyFileToFolder(HttpResponseMessage httpResponse, str
...
+                    secondLineText = GetLine(file, line);
+                }
+                string text = File.ReadAllText(file);
+                text = secondLineText.Length == 0 ? lineToWrite: text.Replace(secondLineText, lineToWrite);
+                File.WriteAllText(file, text);               
+            }
+            else
+                logger.LogInformation(EventIds.ReadMeTextFileIsNotDownloaded.ToEventId(), "Error in downloading readme.txt file for {BatchId}", batchId);
+
+            return true;
```
[the remark this answers is not in the document]
Comment (Avani13709) [reply UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Common/Helpers/FileShareService.cs]: Modified method for returning either true / false based on condition whether file is downloaded or not.

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Webjob.UnitTests/Services/FulfilmentFileShareServiceTest.cs [hunk 12]
```diff
@@ -100,5 +101,58 @@ public void WhenInValidRequest_ThenDownloadFileShareServiceFilesReturnsNoFile()
...
+            bool isFileDownloaded = false;
+            string batchId = "7b4cdf10-adfa-4ed6-b2fe-d1543d8b7272";
+            string exchangeSetRootPath = @"D:\\Downloads";
+            string filePath = "TestFilePath";
+
+            A.CallTo(() => fakefileShareService.DownloadReadMeFile(A<string>.Ignored, A<string>.Ignored, A<string>.Ignored)).Returns(isFileDownloaded);
+            isFileDownloaded = await fulfilmentFileShareService.DownloadReadMeFile(filePath, batchId, exchangeSetRootPath);
+
+            Assert.AreEqual(false, isFileDownloaded);
+        }
```
[the remark this answers is not in the document]
Comment (Avani13709) [reply UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Webjob.UnitTests/Services/FulfilmentFileShareServiceTest.cs:152]: Both test cases will be valid as method has been modified for either true/ false based on condition whether file is downloaded or not.

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Webjob.UnitTests/Services/FulfilmentFileShareServiceTest.cs [hunk 13]
```diff
@@ -100,5 +101,54 @@ public void WhenInValidRequest_ThenDownloadFileShareServiceFilesReturnsNoFile()
...
+            string filePath = "TestFilePath";
+
+            A.CallTo(() => fakefileShareService.DownloadReadMeFile(A<string>.Ignored, A<string>.Ignored, A<string>.Ignored)).Returns(isFileDownloaded);
+            isFileDownloaded = await fulfilmentFileShareService.DownloadReadMeFile(filePath, batchId, exchangeSetRootPath);
+           
+            Assert.AreEqual(true, isFileDownloaded);
+        }
+
+        [Test]
+        public async Task WhenRequestDownloadReadMeFile_ThenReturnsFalseIfFileIsDownloaded()
```
Comment (martyn-fewtrell) [inline UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Webjob.UnitTests/Services/FulfilmentFileShareServiceTest.cs:141]: Thanks - assume this should now be ThenReturnsFalseIfFileIsNotDownloaded

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Webjob.UnitTests/Services/FulfilmentFileShareServiceTest.cs (hunk 12 above)
[the remark this answers is not in the document]
Comment (martyn-fewtrell) [reply UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Webjob.UnitTests/Services/FulfilmentFileShareServiceTest.cs:152]: Thanks Avani

**Review by martyn-fewtrell — COMMENTED** [review]
Thanks both - if we can just update that Test name ill merge to main.

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Webjob.UnitTests/Services/FulfilmentFileShareServiceTest.cs (hunk 13 above)
Comment (Avani13709) [reply UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Webjob.UnitTests/Services/FulfilmentFileShareServiceTest.cs]: Done

**Review by martyn-fewtrell — APPROVED** [review]
Thanks

**Pull request merged**